keras_cv/datasets/pascal_voc/segmentation.py

```diff
@@ -0,0 +1,349 @@
+                    "bbox": [ymin, xmin, ymax, xmax],
+                    "is_truncated": is_truncated,
+                    "is_difficult": is_difficult,
+                }
+            )
+
+        return {"width": width, "height": height, "objects": objects}
+
+
+def _get_image_ids(data_dir, split):
+    data_file_mapping = {"train": "train.txt", "eval": "val.txt", None: "trainval.txt"}
+    with tf.io.gfile.GFile(
+        os.path.join(data_dir, "ImageSets", "Segmentation", data_file_mapping[split]),
+        "r",
+    ) as f:
+        image_ids = f.read().splitlines()
+        logging.info(f"Received {len(image_ids)} images for {split} dataset.")
+        return image_ids
+
+
+def _parse_single_image(image_file_path):
+    data_dir, image_file_name = os.path.split(image_file_path)
+    data_dir = os.path.normpath(os.path.join(data_dir, os.path.pardir))
+    image_id, _ = os.path.splitext(image_file_name)
+    class_segmentation_file_path = os.path.join(
+        data_dir, "SegmentationClass", image_id + ".png"
+    )
+    object_segmentation_file_path = os.path.join(
+        data_dir, "SegmentationObject", image_id + ".png"
+    )
+    annotation_file_path = os.path.join(data_dir, "Annotations", image_id + ".xml")
+    image_annotations = _parse_annotation_data(annotation_file_path)
+
+    result = {
+        "image/filename": image_id + ".jpg",
+        "image/file_path": image_file_path,
+        "segmentation/class/file_path": class_segmentation_file_path,
+        "segmentation/object/file_path": object_segmentation_file_path,
+    }
+    result.update(image_annotations)
+    # Labels field should be same as the 'object.label'
+    labels = list(set([o["label"] for o in result["objects"]]))
+    result["labels"] = sorted(labels)
+    return result
+
+
+def _build_metadata(data_dir, image_ids):
+    # Parallel process all the images.
+    image_file_paths = [
+        os.path.join(data_dir, "JPEGImages", i + ".jpg") for i in image_ids
+    ]
+    pool_size = 10 if len(image_ids) > 10 else len(image_ids)
+    with multiprocessing.Pool(pool_size) as p:
+        metadata = p.map(_parse_single_image, image_file_paths)
+
+    # Transpose the metadata which convert from list of dict to dict of list.
+    keys = [
+        "image/filename",
+        "image/file_path",
+        "segmentation/class/file_path",
+        "segmentation/object/file_path",
+        "labels",
+        "width",
+        "height",
+    ]
+    result = {}
+    for key in keys:
+        values = [value[key] for value in metadata]
+        result[key] = values
+
+    # The ragged objects need some special handling
+    for key in ["label", "pose", "bbox", "is_truncated", "is_difficult"]:
+        values = []
+        objects = [value["objects"] for value in metadata]
+        for object in objects:
+            values.append([o[key] for o in object])
+        result["objects/" + key] = values
+    return result
+
+
+def _decode_png_mask(mask):
```

Comment:
Do we want to jit_compile this?

Reply:
Too late..

Reply:
Opps, submitted too quickly. I don't have the number here. Since usually the dataset load/map happens on CPU, the gain of jit_compile=True might not be very big.

Will run some performance benchmark as a follow up task.

Reply:
Run with following snippet, and the result is like below:
```
============================================
py function result:
warmup_time: 0.016756296157836914 secs
total_time: 3.9043142795562744 secs
============================================
tf function result:
warmup_time: 0.1970975399017334 secs
total_time: 0.6691527366638184 secs
============================================
tf function with xla result:
warmup_time: 0.5804996490478516 secs
total_time: 0.43773317337036133 secs
```


```
import os
import time

import tensorflow as tf

LOCAL_DATA_PATH = '~/pascal_voc_2012/VOCdevkit/VOC2012/'

VOC_COLORMAP = [[0, 0, 0], [128, 0, 0], [0, 128, 0], [128, 128, 0],
                [0, 0, 128], [128, 0, 128], [0, 128, 128], [128, 128, 128],
                [64, 0, 0], [192, 0, 0], [64, 128, 0], [192, 128, 0],
                [64, 0, 128], [192, 0, 128], [64, 128, 128], [192, 128, 128],
                [0, 64, 0], [128, 64, 0], [0, 192, 0], [128, 192, 0],
                [0, 64, 128]]
color_map_to_label = [0] * (256 ** 3)
for i, colormap in enumerate(VOC_COLORMAP):
  color_map_to_label[
    (colormap[0] * 256 + colormap[1]) * 256 + colormap[2]] = i
# There is a special mapping with [224, 224, 192] -> 255
color_map_to_label[224 * 256 * 256 + 224 * 256 + 192] = 255
color_map_to_label = tf.constant(color_map_to_label)


def decode_with_python_function(mask_file_path):
  mask = tf.io.decode_png(tf.io.read_file(mask_file_path))
  mask = tf.cast(mask, tf.int32)
  mask = mask[:, :, 0] * 256 * 256 + mask[:, :, 1] * 256 + mask[:, :, 2]
  mask = tf.expand_dims(tf.gather(color_map_to_label, mask), -1)
  return mask

decode_with_jit_compile = tf.function(func=decode_with_python_function, jit_compile=True)
decode_with_tf_function = tf.function(func=decode_with_python_function)

mask_file_path = os.path.expanduser(LOCAL_DATA_PATH) + 'SegmentationClass/2011_003255.png'


def benchmark_test(fn, mask_file_path, iteration):
  start = time.time()
  # warmup
  fn(mask_file_path)
  warmup_time = time.time() - start
  
  start = time.time()
  for _ in range(iteration):
    fn(mask_file_path)
    
  end = time.time()
  total_time = end - start
  print(f"warmup_time: {warmup_time} secs")
  print(f"total_time: {total_time} secs")


print("============================================")
print("py function result:")
benchmark_test(decode_with_python_function, mask_file_path, 1000)

print("============================================")
print("tf function result:")
benchmark_test(decode_with_tf_function, mask_file_path, 1000)

print("============================================")
print("tf function with xla result:")
benchmark_test(decode_with_jit_compile, mask_file_path, 1000)

```

Reply:
We can ignore the pure python function result, since dataset.map will compile the function into tf.function.

There is about 0.4 sec overhead for XLA compilation, and 1000 step result is 0.2 sec faster. Since 0.4 is just a one time cost, the more image we have, the more gain we will have for using xla.

Reply:
Nice, I suppose that your function, as you in see we have others colormapping type for other segmentation datasets, converted as keras preprocessing layer, will probably let use to preprocess at the batch level on GPU also faster.

Reply:
Also, did a comparison with https://stackoverflow.com/questions/50244982/read-image-in-tensorflow-with-indexed-color-value-for-semantic-segmentation-task/71413414#71413414, that solution is much more slower (90 sec for 1000 records), probably due to match of 1d tensor (eg [128, 0, 0]), instead of encoded value.

Reply:
https://github.com/keras-team/keras-cv/pull/953 for adding the tf.function annotation.

Reply:
Encoding the three pixels with `r*256**2+g*256+b` really is a good idea, but i think your benchmark "cheats" by always using the same sample. That doesn't trigger retracing, while optimizing caching by always using the same data, hence favoring access operations.

I improved the benchmark and ran it multiple times, using the pixel encoding in both strategies, single and multiple samples, with and without `experimental_follow_type_hints`.

I believe my implementation is faster in most cases, while not requiring the allocation of a tensor of shape [256**3].

Benchmark code: https://gist.github.com/lucasdavid/c48d39399b0bbe167cc2f5c056beadf0
<details>
<summary>Benchmark logs</summary>

```shell
(t) ldavid@shepard:~$ CUDA_VISIBLE_DEVICES="" python test_voc_colormap.py --opt a --sampling first --data-dir workspace/datasets/voc/VOCdevkit/VOC2012
[INFO MESSAGES]
Benchmarking opt a sampling:first iterations:1000 follow-type-hints:False
py_fn | warmup:  0.0226 s | total:  4.1935 s
py_fn | warmup:  0.1803 s | total:  0.6522 s
py_fn | warmup:  0.5773 s | total:  0.4188 s
(t) ldavid@shepard:~$ CUDA_VISIBLE_DEVICES="" python test_voc_colormap.py --opt b --sampling first --data-dir workspace/datasets/voc/VOCdevkit/VOC2012
[INFO MESSAGES]
Benchmarking opt b sampling:first iterations:1000 follow-type-hints:False
py_fn | warmup:  0.0238 s | total:  4.8490 s
py_fn | warmup:  0.2218 s | total:  0.2861 s
py_fn | warmup:  0.0594 s | total:  0.2358 s
(t) ldavid@shepard:~$ CUDA_VISIBLE_DEVICES="" python test_voc_colormap.py --opt b --sampling first --data-dir workspace/datasets/voc/VOCdevkit/VOC2012
[INFO MESSAGES]
Benchmarking opt b sampling:first iterations:1000 follow-type-hints:False
py_fn | warmup:  0.0229 s | total:  4.8222 s
py_fn | warmup:  0.2196 s | total:  0.2447 s
py_fn | warmup:  0.0623 s | total:  0.2472 s
(t) ldavid@shepard:~$ CUDA_VISIBLE_DEVICES="" python test_voc_colormap.py --opt a --sampling first --data-dir workspace/datasets/voc/VOCdevkit/VOC2012
[INFO MESSAGES]
Benchmarking opt a sampling:first iterations:1000 follow-type-hints:False
py_fn | warmup:  0.0209 s | total:  4.1461 s
py_fn | warmup:  0.1806 s | total:  0.6624 s
py_fn | warmup:  0.5778 s | total:  0.4153 s
(t) ldavid@shepard:~$ CUDA_VISIBLE_DEVICES="" python test_voc_colormap.py --opt a --sampling first --data-dir workspace/datasets/voc/VOCdevkit/VOC2012 --experimental-follow-type-hints
[INFO MESSAGES]
Benchmarking opt a sampling:first iterations:1000 follow-type-hints:True
py_fn | warmup:  0.0223 s | total:  4.1025 s
py_fn | warmup:  0.1827 s | total:  4.1046 s
(t) ldavid@shepard:~$ CUDA_VISIBLE_DEVICES="" python test_voc_colormap.py --opt b --sampling first --data-dir workspace/datasets/voc/VOCdevkit/VOC2012 --experimental-follow-type-hints
[INFO MESSAGES]
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
Benchmarking opt b sampling:first iterations:1000 follow-type-hints:True
py_fn | warmup:  0.0229 s | total:  4.8926 s
py_fn | warmup:  0.2413 s | total:  3.6635 s
(t) ldavid@shepard:~$ CUDA_VISIBLE_DEVICES="" python test_voc_colormap.py --opt a --sampling first --data-dir workspace/datasets/voc/VOCdevkit/VOC2012 --experimental-follow-type-hints
[INFO MESSAGES]
Benchmarking opt a sampling:first iterations:1000 follow-type-hints:True
py_fn | warmup:  0.0215 s | total:  4.1227 s
py_fn | warmup:  0.1848 s | total:  4.1191 s
(t) ldavid@shepard:~$ CUDA_VISIBLE_DEVICES="" python test_voc_colormap.py --opt b --sampling first --data-dir workspace/datasets/voc/VOCdevkit/VOC2012 --experimental-follow-type-hints
[INFO MESSAGES]
Benchmarking opt b sampling:first iterations:1000 follow-type-hints:True
py_fn | warmup:  0.0224 s | total:  4.9606 s
py_fn | warmup:  0.2549 s | total:  3.7402 s
(t) ldavid@shepard:~$ CUDA_VISIBLE_DEVICES="" python test_voc_colormap.py --opt a --sampling many --data-dir workspace/datasets/voc/VOCdevkit/VOC2012 --experimental-follow-type-hints
[INFO MESSAGES]
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
Benchmarking opt a sampling:many iterations:1000 follow-type-hints:True
py_fn | warmup:  0.0231 s | total:  4.1516 s
py_fn | warmup:  0.1907 s | total:  4.0629 s
(t) ldavid@shepard:~$ CUDA_VISIBLE_DEVICES="" python test_voc_colormap.py --opt b --sampling many --data-dir workspace/datasets/voc/VOCdevkit/VOC2012 --experimental-follow-type-hints
[INFO MESSAGES]
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
Benchmarking opt b sampling:many iterations:1000 follow-type-hints:True
py_fn | warmup:  0.0233 s | total:  4.8790 s
py_fn | warmup:  0.2338 s | total:  3.8315 s
(t) ldavid@shepard:~$ CUDA_VISIBLE_DEVICES="" python test_voc_colormap.py --opt b --sampling many --data-dir workspace/datasets/voc/VOCdevkit/VOC2012 --experimental-follow-type-hints
[INFO MESSAGES]
Benchmarking opt b sampling:many iterations:1000 follow-type-hints:True
py_fn | warmup:  0.0224 s | total:  4.8583 s
py_fn | warmup:  0.2371 s | total:  3.8537 s
(t) ldavid@shepard:~$ CUDA_VISIBLE_DEVICES="" python test_voc_colormap.py --opt a --sampling many --data-dir workspace/datasets/voc/VOCdevkit/VOC2012 --experimental-follow-type-hints
[INFO MESSAGES]
Benchmarking opt a sampling:many iterations:1000 follow-type-hints:True
py_fn | warmup:  0.0208 s | total:  4.2101 s
py_fn | warmup:  0.1872 s | total:  3.9629 s
```
</details>

Reply:
Thanks, note that `experimental_follow_type_hint` is formally deprecated.

It is also important that the function could work directly in batches. 
See:
https://github.com/keras-team/keras-cv/pull/953#issuecomment-1287003209

Reply:
Well, the current implementation in main branch isn't batched, is it? I guess it doesn't matter anymore, but for honesty purposes, I [re-implemented the benchmark](https://gist.github.com/lucasdavid/c48d39399b0bbe167cc2f5c056beadf0) using batches and without experimental_follow_type_hint.

qlzh727's implementation is indeed faster when batching (~3x). Still, it has the drawback of huge memory footprint (64MB) for storing all color codes in a (256**3)-sized array.

```shell
(t) ldavid@shepard:~$ CUDA_VISIBLE_DEVICES="" python test_voc_colormap.py --opt a --data-dir workspace/datasets/voc/VOCdevkit/VOC2012
Benchmarking opt a iterations:1000
mask_decode_py  | warmup:  0.0704 s | total:  0.0642 s
mask_decode_tf  | warmup:  0.0851 s | total:  0.0469 s
mask_decode_jit | warmup:  0.0274 s | total:  0.0153 s
(t) ldavid@shepard:~$ CUDA_VISIBLE_DEVICES="" python test_voc_colormap.py --opt b --data-dir workspace/datasets/voc/VOCdevkit/VOC2012
Benchmarking opt b iterations:1000
mask_decode_py  | warmup:  0.1805 s | total:  0.1733 s
mask_decode_tf  | warmup:  0.2060 s | total:  0.1662 s
mask_decode_jit | warmup:  0.3092 s | total:  0.3028 s
```